### PR TITLE
EG-644-1, Handle loss of sessionId after timeout using temporary EG_NEW_SESSIONID session key

### PR DESCRIPTION
**Pull request by cmrhmrc**

### CODE REVIEW

File: app/controllers/actions/SessionIdAction.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import uk.gov.hmrc.http.SessionKeys
+import play.api.Logger
+import scala.concurrent.{ExecutionContext, Future}
+
+trait SessionIdAction extends ActionBuilder[Request, AnyContent] with ActionFunction[Request, Request] {
+  val EgNewSessionIdName: String = "EG_NEW_SESSIONID"
+}
+
+class SessionIdActionImpl @Inject()(val parser: BodyParsers.Default)
+                               (implicit val executionContext: ExecutionContext)
```
Comment (woollett-r) [inline app/controllers/actions/SessionIdAction.scala:32]: IntelliJ reports trailing whitespace

File: app/controllers/actions/SessionIdAction.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import javax.inject.Inject
+import play.api.mvc.request.Cell
+import play.api.mvc.request.RequestAttrKey
+import play.api.mvc._
+import uk.gov.hmrc.http.SessionKeys
+import play.api.Logger
+import scala.concurrent.{ExecutionContext, Future}
+
+trait SessionIdAction extends ActionBuilder[Request, AnyContent] with ActionFunction[Request, Request] {
+  val EgNewSessionIdName: String = "EG_NEW_SESSIONID"
```
Comment (woollett-r) [inline app/controllers/actions/SessionIdAction.scala:28]: I'd start with a lower case "e"; this isn't C# you know.

File: app/controllers/actions/SessionIdAction.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import javax.inject.Inject
+import play.api.mvc.request.Cell
+import play.api.mvc.request.RequestAttrKey
+import play.api.mvc._
+import uk.gov.hmrc.http.SessionKeys
+import play.api.Logger
+import scala.concurrent.{ExecutionContext, Future}
+
+trait SessionIdAction extends ActionBuilder[Request, AnyContent] with ActionFunction[Request, Request] {
```
Comment (woollett-r) [inline app/controllers/actions/SessionIdAction.scala:27]: I would add a comment explaining why the action is necessary. Someone might appreciate it one day.

File: app/controllers/actions/SessionIdAction.scala (hunk 2 above)
Comment (cmrhmrc) [reply app/controllers/actions/SessionIdAction.scala:28]: Its Scala :) https://docs.scala-lang.org/style/naming-conventions.html

File: app/controllers/actions/SessionIdAction.scala (hunk 1 above)
Comment (cmrhmrc) [reply app/controllers/actions/SessionIdAction.scala:32]: will remove

File: app/controllers/actions/SessionIdAction.scala (hunk 3 above)
Comment (cmrhmrc) [reply app/controllers/actions/SessionIdAction.scala:27]: will add

File: app/controllers/actions/SessionIdAction.scala (hunk 2 above)
Comment (cmrhmrc) [reply app/controllers/actions/SessionIdAction.scala:28]: They call it upper camel case i.e. camel case beginning with an initial upper case char

File: app/controllers/GuidanceController.scala [hunk 4]
```diff
@@ -133,34 +132,43 @@ class GuidanceController @Inject() (
...
       logger.error(s"Session Id missing from request when required")
       Future.successful(Left(BadRequestError): RequestOutcome[T])
     } { sessionId =>
-      logger.info(s"withSession, EG_SESSIONID ${request.session.get("EG_SESSION")}, sessionId $sessionId")
+      logger.info(s"Found existing sessionId = $sessionId")
       block(sessionId.value)
     }
 
-  private def retreiveCacheAndRedirectToProcess(id: String, sessionId: String, processStartUrl: (String, String) => Future[RequestOutcome[String]])(
+  private def retreiveCacheAndRedirectToView(id: String, retreiveAndCache: (String, String) => Future[RequestOutcome[String]])(
```
Comment (woollett-r) [inline app/controllers/GuidanceController.scala:148]: Sorry to be Mr Pedantic, but retrieve is "i" before "e".

Comment (cmrhmrc) [reply app/controllers/GuidanceController.scala:148]: the shame :( you are right of course, will change

File: app/controllers/GuidanceController.scala [hunk 5]
```diff
@@ -133,34 +132,43 @@ class GuidanceController @Inject() (
...
       case Left(NotFoundError) =>
         logger.warn(s"Unable to find process $id and render using sessionId $sessionId")
         NotFound(errorHandler.notFoundTemplate)
       case Left(err) =>
         logger.error(s"Error $err returned from Guidance service when trying to access process $id using sessionId $sessionId")
         InternalServerError(errorHandler.internalServerErrorTemplate)
     }
+  }
+
+  private def existingOrNewSessionId()(implicit request: Request[_]): (String, Option[String]) =
```
Comment (woollett-r) [inline app/controllers/GuidanceController.scala:167]: IntelliJ is reporting trailing white space at the end of the line.

Comment (cmrhmrc) [reply app/controllers/GuidanceController.scala:167]: will remove

File: test/controllers/actions/SessionIdActionSpec.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+      val result: Future[Result] = target{ request => Results.Ok.withSession(request.session)}
+                                    .onPageLoad()(FakeRequest("GET", path).withSession(SessionKeys.sessionId -> sessionId))
+      session(result).get(SessionKeys.sessionId) mustBe Some(sessionId)
+      status(result) mustBe Status.OK
+    }
+
+    "Convert EG_NEW_SESSIONID into a sessionId" in new Test {
+
+      target{ request =>
```
Comment (woollett-r) [inline test/controllers/actions/SessionIdActionSpec.scala:58]: IntelliJ reports whitespace at end of line.

File: test/controllers/actions/SessionIdActionSpec.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        request.session.data.get(sessionIdAction.EgNewSessionIdName) mustBe None
+        request.session.data.get(SessionKeys.sessionId) mustBe Some(egSessionId)
+        Results.Ok.withSession(request.session)
+      }.onPageLoad()(FakeRequest("GET", path).withSession(sessionIdAction.EgNewSessionIdName -> egSessionId))
+      
+    }
+
+    "Add nothing if neither sessionId or EG_NEW_SESSIONID found" in new Test {
+
+      target{ request =>
```
Comment (woollett-r) [inline test/controllers/actions/SessionIdActionSpec.scala:68]: IntelliJ reports whitespace at end of line

File: test/controllers/actions/SessionIdActionSpec.scala (hunk 6 above)
Comment (cmrhmrc) [reply test/controllers/actions/SessionIdActionSpec.scala:58]: will remove

File: test/controllers/actions/SessionIdActionSpec.scala (hunk 7 above)
Comment (cmrhmrc) [reply test/controllers/actions/SessionIdActionSpec.scala:68]: will remove

**Pull request merged**